src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs

```diff
@@ -185,7 +220,53 @@ IEnumerator IEnumerable.GetEnumerator()
                 return GetEnumerator();
             }
         }
+
+        private class AdditionalActivityItemsLogScope<TValue> : IEnumerable<KeyValuePair<string, TValue>>
```

Comment:
>         [](start = 0, length = 8)

We can have this class only for Baggage. For Tags, we still can implement ToString on the TagsLinkedList and that will save us the object allocation. 

For this class, also we can create it using the Activity object and not IEnumerable<KeyValuePair<string, TValue?>>. If we do that, we'll ensure we can always report the updated Baggage if it is changed between calls. Also inside this object we can cache the StringBuilder object if it is created first time. so we don't have to reallocate it every time ToString is called.
We can optimize more if we need to by trying to allocate a buffer to some specific size on the stack (say 500 character length) and we can use this buffer to fill in till we hit the max and then we reallocate at that time. but let's not start with that now and look how it goes first without this optimization.

Reply:
> We can have this class only for Baggage. For Tags, we still can implement ToString on the TagsLinkedList and that will save us the object allocation.

Will have a look at this. 

> For this class, also we can create it using the Activity object and not IEnumerable<KeyValuePair<string, TValue?>>. If we do that, we'll ensure we can always report the updated Baggage if it is changed between calls.

See my comment above. We do always enumerate the current items (expect I missed something).

> Also inside this object we can cache the StringBuilder object if it is created first time. so we don't have to reallocate it every time ToString is called.

Sorry. I don't understand where you want to cache the SB. We could have it not in ToString but at instance level but thats not thread-safe (and I guess it should be or not?).

Reply:
> Sorry. I don't understand where you want to cache the SB

I meant to cache it in `AdditionalActivityItemsLogScope`. For thread safety, it is very unlikely 2 or more threads will use the same `AdditionalActivityItemsLogScope` object in same time, so in `AdditionalActivityItemsLogScope.ToString()` would be ok to do `lock(_sb)` as it would be a cheap operation where _sb is the StringBuilder objected created first time ToString() is called.

Reply:
Done:
- Cache `StringBuilder` in wrapper object (with lock on the sb)
- Overwrite `TagsLinkedList.ToString()` and use this as Scope rather than a wrapper object